src/grammar_optimiser/enumerate_subtrees.jl

```diff
@@ -0,0 +1,102 @@
+using DataStructures
+
+"""
+    enumerate_subtrees(tree::RuleNode, g::AbstractGrammar)
+
+Enumerates all subtrees of a given tree.
+# Arguments
+- `tree::RuleNode`: the tree to enumerate the subtrees of
+- `g::AbstractGrammar`: the grammar to use
+# Result
+- `subtrees::Vector{RuleNode}`: a list of all subtrees of the tree
+"""
+function enumerate_subtrees(tree::RuleNode, g::AbstractGrammar)
```

Comment:
Maybe you could make this an iterator instead? So you don't have to hold every possible sub-tree in memory all the time. 
Also you can then `collect(iterator)` to get the same list.